### ISSUE TITLE
Charst problem

### ISSUE THREAD

**Issue by puschie286** [wkhtmltopdf/wkhtmltopdf#4951]
**wkhtmltopdf version(s) affected**: 0.12.6

**OS information**
Windows 10

**Description**
german special characters are rendered as ¿½ or �

**How to reproduce**
use "<head><meta charset = \"ISO-8859-1\"></head>Ü Ü ü ö ö ö test"
or "<head><meta charset = \"utf-8\"></head> Ü Ü ü ö ö ö test"

**Expected behavior**
PDF should contain: Ü Ü ü ö ö ö test

**Comment by PhilterPaper** [comment]
The first thing to check would be to be absolutely positively certain that your HTML file was actually edited in Latin-1 (or Windows-1252) when you claim "ISO-8859-1" encoding, **OR** was actually edited in UTF-8 for claimed "UTF-8". I'm not sure what's going on with the '¿½' (the two bytes in Latin-1 don't look like a valid UTF-8 character, unless they are actually a _pair_ of UTF-8 characters themselves -- can you look at the raw bytes?). The ?-in-black-diamond says that it's an invalid UTF-8 character (something x80 or higher that doesn't meet the encoding system). You should find an editor or dump that you can actually look at the hex values and confirm that they are either Latin-1 or UTF-8 encoding for your HTML.

**Comment by puschie286** [comment]
so you get the correct ouput when using the test string ?

Well im using visual studio 2017 for writing the string but its defined as char array.
wkhtmltopdf_add_object has a char pointer parameter not wchar, so i pass the string as char array.

maybe thats the problem, i didnt find anything to multibyte handling in the documentation

**Comment by puschie286** [comment]
Here is the full code:
```
// init
int result = wkhtmltopdf_init( 0 );
// global config
wkhtmltopdf_global_settings* settings = wkhtmltopdf_create_global_settings();
wkhtmltopdf_set_global_setting( settings, "out", "D:\\test.pdf" );
// converter
wkhtmltopdf_converter* converter = wkhtmltopdf_create_converter( settings );
// page config
wkhtmltopdf_object_settings* page = wkhtmltopdf_create_object_settings();
std::string htmlPage = "<head><meta charset = \"utf-8\"></head> Ü Ü ü ö ö ö test";
// add page to conv
wkhtmltopdf_add_object( converter, page, htmlPage.c_str() );
// convert
wkhtmltopdf_convert( converter );
// cleanup
wkhtmltopdf_destroy_converter( converter );
```
and the result pdf : [test.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/5958836/test.pdf)

**Comment by PhilterPaper** [comment]
I'll bet a full steak dinner that when you created this file to build the HTML and run wkHTMLtoPDF, you created it in something other than UTF-8 (probably Latin-1 or Windows-1252). Your accented characters &Uuml; etc. have to be UTF-8 in the HTML source. They will not be automagically converted to UTF-8 if you typed them in as a single-byte source and declared the charset to be UTF-8. You either need to tell your editor to edit and save in UTF-8, or run some kind of file translator/converter to change your Latin-1 source file into UTF-8.

See the row of ?-in-black-diamonds? That means that the characters are out of ASCII range, yet not valid UTF-8 codes. Most likely they are in the range x80-xFF as single byte characters.

**Comment by puschie286** [comment]
sorry but i dont get what you are trying to say

there is no html - its just the c string ( as char array ).
normally i would use a wchar for multibyte text but wkhtmltopdf_add_object only accept char

**Comment by puschie286** [comment]
updated the question to make it easier to understand.

**Comment by PhilterPaper** [comment]
'ï¿½' are the Latin-1 bytes that make up the UTF-8 invalid character mark ?-in-black-diamond (U+FFFD). You're mixing UTF-8 with Latin-1 usage. You need to keep _everything_ in one encoding or the other from the start, and not mix them (e.g., copy and paste output from one charset's run into the other).

It doesn't matter that this is C code -- it's still seen as HTML by wkHTMLtoPDF when it's read in. wchar (wide characters) might have helped here, but if it can't be accepted, you'll have to stay with a single byte encoding. If you stick with a single byte encoding such as Latin-1, ISO-8859-1, or even Windows-1252 from beginning to end (editing your files, charset passed to wkHTMLtoPDF, etc.) it should work. I can see that you're constantly mixing UTF-8 charset with Latin-1 text, or vice-versa, and that's not going to work.

**Comment by puschie286** [comment]
everything is in one encoding on my side - windows-1252 doesnt work for me with wkHTMLtoPDF - thats the whole point

thank you for your effort, but we do not get anywhere like this.
please check the example code and verify if this works for you or not

**Comment by PhilterPaper** [comment]
wkHTMLtoPDF will support many encodings. My point was if you are going to declare to wkHTMLtoPDF that your input encoding (charset) is UTF-8, that you _have_ to supply your text (HTML content) in UTF-8 encoded characters, not Latin-1 etc. ASCII text works in any encoding, but accented characters as you're trying are two bytes each in UTF-8, and only one byte each in Latin-1.

**Comment by puschie286** [comment]
of cause the input encoding and the supplied text ( or html how you call it ) are in the same encoding - everything else wouldn't make any sense.
i know that this wasnt clear - thats why i added my default encoding to the description and title.

**Comment by PhilterPaper** [comment]
The only other thing that comes to mind is that something in the food chain there (most likely the C compiler itself) is "helpfully" changing UTF-8 text to Latin-1, 'cuz that's all it knows how to deal with (especially if you're not using **wchar** or the like). So, you may be carefully editing in UTF-8, but those accented characters are being changed to Latin-1 (or worse, gibberish) somewhere before wkHTMLtoPDF gets them.

Try doing everything in Windows-1252 or ISO-8859-1 (_no_ copy-paste unless you're sure your source is a matching encoding), including entry of all the accented characters _and_ the declared charset. If _that_ doesn't work, there are deeper problems here. If it does work, change your C until you get the compiler to accept multibyte (UTF-8) characters.

**Comment by puschie286** [comment]
checked everything 10 times - the source is 100% windows-1252 ( database collation is Latin1_General_CI_AS, where the data is stored ).

my current workaround is to convert it to Unicode and then to utf8 ( and display with utf8 charset )

**Comment by PhilterPaper** [comment]
Did the data go into the database **in the first place** as Latin-1? If you shoved UTF-8 text into a Latin-1 field, it doesn't magically get converted! You should see multibyte trash in something like phpMyAdmin. In your sample HTML, _how_ did you create the accented characters (e.g., U-umlaut)? You're absolutely positively sure they're single-byte Latin-1 or Windows-1252? If they came out of a database, or were copy-pasted from somewhere else, do you have proof that they're _not_ actually UTF-8?

**Comment by puschie286** [comment]
Yes, it’s 100% Latin-1 text - we don’t have any problems within other software components, all text is printed correctly as html or cmd output ( and yes we use the Windows charset everywhere )
The example code is strings are created/converted with notepad++ so I’m pretty sure my test used the correct charset
